shcomplete/tfdf.py

```diff
@@ -0,0 +1,112 @@
+import logging
+import os
+import re
+from math import log
+from clint.textui import progress
+
+import pygtrie
+import numpy
+
+
+def get_tries(args, _log):
+    """
+    Return a list of tries, one for each file where the keys are the command lines.
+    """
+    tries = []
+    max_len = args.max_length
+    for root, _, files in os.walk(args.directory):
+        for file_name in files:
+            path_to_file = os.path.join(root, file_name)
+            _log.info("parsing %s", path_to_file)
+            trie = pygtrie.StringTrie(separator=" ")
+            tries.append((path_to_file, trie))
+            with open(path_to_file, "r") as f:
+                lines = f.readlines()
+                for line in lines:
+                    line = [l for l in line.rstrip().replace("\t", " ").split(" ") if l][:max_len]
+                    key = " ".join(line)
+                    if key not in trie:
+                        trie[key] = 0
+                    prefixes = []
+                    for i in range(1, len(line)+1):
```

Comment:
Quadratic complexity. Rewrite to single assignment of the leaf + tree traversal with value updates (actually the code I wrote some time ago)